### PR TITLE
feat: introduce macos support

### PR DESCRIPTION
**Pull request by oowl**
recreate https://github.com/Sherlock-Holo/fuse3/pull/90

### CODE REVIEW

**Comment by oowl** [comment]
If you want to test fuse3 macOS support, you can check https://github.com/apache/opendal/pull/5136#issuecomment-2367398541 and try to use it.

**Comment by oowl** [comment]
also cc @radumarias, Sorry for the long long long time delay.

**Comment by radumarias** [comment]
> also cc @radumarias, Sorry for the long long long time delay.

no worries man :) thank you for getting so far :) will read about that and give it a try, eager for this :)

File: src/path/inode_path_bridge.rs [hunk 1]
```diff
@@ -997,7 +997,7 @@ where
             .copied()
             .filter_map(|inode| inode_name_manager.get_absolute_path(inode))
             .collect::<Vec<_>>();
-        let paths = paths.iter().map(|path| path.as_ref()).collect::<Vec<_>>();
+        let paths = paths.iter().map(|path: &PathBuf| path.as_ref()).collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (oowl) [reply src/path/inode_path_bridge.rs]: I forgot to delete it, also I can not remember why I added it.

File: src/raw/connection/tokio.rs [hunk 2]
```diff
@@ -191,6 +220,114 @@ impl BlockFuseConnection {
...
+            debug!("mount_thread start");
+            let fd0 = sock0.as_raw_fd();
+            let mut binding = Command::new(binary_path);
+            let child = binding
+                .env(ENV, fd0.to_string())
+                .env("_FUSE_CALL_BY_LIB", "1")
+                .env("_FUSE_COMMVERS", "2")
+                .env("_FUSE_DAEMON_PATH", exec_path)
+                .args(vec![ options, mount_path]);
+            let child = child.spawn()?.wait_with_output().await?;
```
[the remark this answers is not in the document]
Comment (oowl) [reply src/raw/connection/tokio.rs:269]: From my side, `wait_with_output` returns `Result`, return `OK` is not mean execution success or failure, we need to know what is exited from the real command.

File: src/lib.rs [hunk 3]
```diff
@@ -214,3 +237,15 @@ fn find_fusermount3() -> io::Result<PathBuf> {
...
     })
 }
+
+#[cfg(target_os = "macos")]
+fn find_macfuse_mount() -> io::Result<PathBuf> {
+    if Path::new("/Library/Filesystems/macfuse.fs/Contents/Resources/mount_macfuse").exists() {
+        Ok(PathBuf::from("/Library/Filesystems/macfuse.fs/Contents/Resources/mount_macfuse"))
+    } else {
+        Err(io::Error::new(
+            ErrorKind::Other,
```
[the remark this answers is not in the document]
Comment (oowl) [reply src/lib.rs]: done

File: src/mount_options.rs [hunk 4]
```diff
@@ -244,6 +243,30 @@ impl MountOptions {
         options
     }
 
+    #[cfg(target_os = "macos")]
+    pub(crate) fn build(&self) -> OsString {
+        let mut opts = vec![
+            format!("-o fsname=ofs"),
```
[the remark this answers is not in the document]
Comment (oowl) [reply src/mount_options.rs]: done

File: src/raw/connection/tokio.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sherlock-Holo) [reply src/raw/connection/tokio.rs:269]: use `wait` then check the `ExitStatus` can also do this job

[the remark this answers is not in the document]
Comment (oowl) [reply src/raw/connection/tokio.rs]: done

**Pull request merged**